Enable weak constraints

AntoniusW opened the pull request:
Adds weak constraints to Alpha. The parser is able to read them and in their presence the solver switches to a branch-and-bound algorithm. Returned answer sets are adorned with a valuation (weights at different levels as specified by the weak constraints).

AntoniusW committed:
Add parsing and initial rewriting of weak constraints.

- Refined grammar file.
- ParseTreeVisitor creates rules with WeakConstraintAtom in head.
- New WeakConstraint type representing parsed weak constraints.
- New WeakConstraintAtom as special head atom of weak constraints.
- New WeakConstraintNormalization transforming WeakConstraint rules into
  normal rules with WeakConstraintAtom in head.
- Removed dependence on Literals from IntervalTermToIntervalAtom and
  NaiveGrounder where Atoms are sufficient.

AntoniusW committed:
Merge branch 'master' into weak_constraints

AntoniusW committed:
Refactor callback management for atoms in solver, grounder reports weak
constraints to solver.

- Introduce AtomCallbackManager.
- Move callback functionality out of ChoiceManager.
- Assignment and TrailAssignment work with AtomCallbackManager.
- Add WeakConstraintsManager for weak constraint callbacks.
- Callbacks for choice atoms now handled in ChoicePoint directly.
- NoGoodGenerator generates nogoods representing weak constraints.
- Solver relays weak constraint information to WeakConstraintsManager.

AntoniusW committed:
Enable weak constraints handling in solver; add containsWeakConstraints
to internal Program.

- Add containsWeakConstraints to InternalProgram, and modify
  AnalyzedProgram, Grounder, NaiveGrounder, StratifiedEvaluation, Alpha,
  ChoiceGrounder, DummyGrounder accordingly.
- Introduce WeightedAnswerSet, adapt BasicAnswerSet, DefaultSolver, Main
  to print its valuation.
- Fix bugs in NoGoodGenerator and AtomCallbackManager.
- DefaultSolver switches to branch-and-bound if weak constraints are in
  input program.
- WeakConstraintsManager maintains valuation of current and best known
  assignment/answer set.
- Fix issue #272 in DefaultSolver.

AntoniusW committed:
Polish some comments.

AntoniusW committed:
Merge branch 'master' into weak_constraints

# Conflicts:
#	src/main/java/at/ac/tuwien/kr/alpha/grounder/NaiveGrounder.java

AntoniusW committed:
Satisfy checkstyle and printing polished.

AntoniusW committed:
Add testing facilities for weighted answer sets.

- AnswerSet.compareTo checks for same class.
- WeightedAnswerSet implements compareTo, hashCode, and equals.
- WeakConstraintsManager fix possible NPE.
- TestUtils able to check that answer sets contain an optimum answer.
- WeakConstraintsTests created with some simple integration tests.

AntoniusW committed:
Extend weighted answer sets to allow negative levels.

- WeightedAnswerSet employs TreeMap for weights at levels.
- Add WeightedAnswerSetTest unit tests.
- Refactor TestUtils, add helper to construct WeightedAnswerSet.

AntoniusW committed:
Enable negative levels in optimisation, fix ordering of levels.

- WeakConstraintsManager adapted for correct ordering and negative
  levels.
- Added more tests.

AntoniusW committed:
Increase test timeouts as ChoiceManager has internal checks enabled now.

AntoniusW committed:
Fix bug in WeakConstraintsManager, polish WeightedAnswerSet.

AntoniusW committed:
Refactor WeakConstraintsManager internals.

AntoniusW committed:
Move WeakConstraintsManager to subpackage, adapt necessary visibilities.

AntoniusW committed:
Refactor WeakConstraintsManager (move to own package, split into three classes).

AntoniusW committed:
Bugfix: produce weak constraint association only once per level+terms.

AntoniusW committed:
Bugfix: update weak constraints information at the right time (after growing and before propagating).

AntoniusW committed:
Refactor WeightAtLevelsManager, WeakConstraintsManager, and fix several bugs. Also adding more tests.

- WeakConstraintAtomCallback add toString().
- WeightAtLevelsManager only deals with weights and levels, and no longer needs to know about weak constraints.
- WeakConstraintsManager keeps track of true atoms above level of WeightAtLevelsManager and keeps weak constraints away from WeightAtLevelsManager.
- WeightAtLevelsManagerTest added.
- WeakConstraintsTests comes with significantly more tests.

AntoniusW committed:
Merge branch 'master' into weak_constraints

# Conflicts:
#	src/main/java/at/ac/tuwien/kr/alpha/solver/DefaultSolver.java
#	src/test/java/at/ac/tuwien/kr/alpha/solver/OmigaBenchmarksTest.java
#	src/test/java/at/ac/tuwien/kr/alpha/test/util/TestUtils.java

AntoniusW committed:
Fix merge, adapt tests to junit5.

AntoniusW commented:
This PR should be ready to be reviewed now.

AntoniusW committed:
Fix failing tests.

- WeightAtLevelsManager corrected logging, tolerates 0 weight at level 0.
- WeightAtLevelsManagerTest runs with checks now, added more tests.
- WeakConstraintsTests ignored for naive solver as that one does not optimize.

AntoniusW commented:
Now also with test passing. :-)

madmike200590 commented:
I'm not yet through with the complete review, but I think I found a bug in one my tests:
The following program only has 4 answer sets with weak constraints:
```
{a; b; c}.

:~ a. [1]
:~ b. [2]
:~ c. [3]
```
Alpha output is:
```
michael@michael-laptop:~/asp_snippets/weak_constraints$ java -jar Alpha-bundled.jar -i wc-test.asp
  376 I lections.Reflections Reflections took 98 ms to scan 1 urls, producing 1 keys and 8 values  
Answer set 1:
{ a, c }
Optimization: [4@0]
Answer set 2:
{ c }
Optimization: [3@0]
Answer set 3:
{ a }
Optimization: [1@0]
Answer set 4:
{  }
Optimization: []
SATISFIABLE
```

Since I didn't specify any limit on answer sets computed, I'd expect to get all of them with their associated weigths. Is that a misunderstanding on my part?

The program without weak constraints, i.e. `{a; b; c}.` yields the expected 8 answer sets:
```
michael@michael-laptop:~/asp_snippets/weak_constraints$ java -jar Alpha-bundled.jar -i wc-test.asp
  399 I lections.Reflections Reflections took 110 ms to scan 1 urls, producing 1 keys and 8 values  
Answer set 1:
{ a, c }
Answer set 2:
{ c }
Answer set 3:
{ a }
Answer set 4:
{  }
Answer set 5:
{ a, b, c }
Answer set 6:
{ a, b }
Answer set 7:
{ b, c }
Answer set 8:
{ b }
SATISFIABLE
```
I haven't done any detailed analysis on this yet, but wanted to raise it anyway so you can take a look as well.

AntoniusW commented:
To me that looks perfectly fine. The solver sees the weak constraints and switches over to optimisation using a branch-and-bound algorithm. That means branches whose valuation is higher than the currently-best-known valuation are cut off immediately and not expanded. So one usually expects a sequence of answer-sets with increasingly better valuation (i.e., less violations). Since the last answer-set has zero violated weak constraints (hence a valuation of 0 at all levels), it is the best answer-set and the solver stops there.

Note that this behaviour is similar to what other ASP solvers do. Enumerating all answer-sets is usually infeasible when optimization is enabled and by using optimization (i.e., weak constraints) one tells the solver that one is interested in good answers only.

AntoniusW committed:
Merge branch 'master' into weak_constraints

# Conflicts:
#	src/main/java/at/ac/tuwien/kr/alpha/grounder/NaiveGrounder.java
#	src/main/java/at/ac/tuwien/kr/alpha/grounder/transformation/IntervalTermToIntervalAtom.java
#	src/main/java/at/ac/tuwien/kr/alpha/grounder/transformation/NormalizeProgramTransformation.java
#	src/test/java/at/ac/tuwien/kr/alpha/test/util/TestUtils.java

AntoniusW committed:
Fix merge issues with aggregate rewriting.

- WeakConstraint fixed toString.
- AggregateOperatorNormalization returns WeakConstraint if rule is of that type.
- IntervalTermToIntervalAtom rewriting of head no longer needs Literals.

madmike200590 commented:
> To me that looks perfectly fine. The solver sees the weak constraints and switches over to optimisation using a branch-and-bound algorithm. That means branches whose valuation is higher than the currently-best-known valuation are cut off immediately and not expanded. So one usually expects a sequence of answer-sets with increasingly better valuation (i.e., less violations). Since the last answer-set has zero violated weak constraints (hence a valuation of 0 at all levels), it is the best answer-set and the solver stops there.
> 
> Note that this behaviour is similar to what other ASP solvers do. Enumerating all answer-sets is usually infeasible when optimization is enabled and by using optimization (i.e., weak constraints) one tells the solver that one is interested in good answers only.

As per our discussion last week, I think we need the following changes:
- default behavior when Alpha is called without a specific number of answer sets to compute: Same as for programs without weak constraints, this should enumerate all answer sets along with their respective weights (needed for development/debugging)
- manual initial bound for answer set search: Add a commandline flag "-mw"/"--maxWeight" that, when used in conjuction with "-n" (number of answer sets) instructs Alpha to find n answer sets where no answer set must exceed the given weight (i.e. use the value of mw as initial bound for branch&bound)

AntoniusW committed:
Merge branch 'master' into weak_constraints

# Conflicts:
#	src/main/java/at/ac/tuwien/kr/alpha/solver/DefaultSolver.java

AntoniusW committed:
Introduce OptimizingSolver for weak constraints, SolverFactory returns OptimizingSolver if weak constraints are present.

- OptimizingSolver extends DefaultSolver, implements branch-and-bound search.
- DefaultSolver visibility of methods/fields adapted for inheritance.
- AnswerSet provides getPredicateInstances for construction of WeightedAnswerSet from BasicAnswerSet.
- Fix logging level in WeakConstraintsManager and WeakConstraintsTests.

AntoniusW committed:
Add some javadoc to OptimizingSolver.

madmike200590 commented:
> As per our discussion last week, I think we need the following changes:
> 
> * default behavior when Alpha is called without a specific number of answer sets to compute: Same as for programs without weak constraints, this should enumerate all answer sets along with their respective weights (needed for development/debugging)
> * manual initial bound for answer set search: Add a commandline flag "-mw"/"--maxWeight" that, when used in conjuction with "-n" (number of answer sets) instructs Alpha to find n answer sets where no answer set must exceed the given weight (i.e. use the value of mw as initial bound for branch&bound)
> * bugfix: When Alpha finds an optimum solution, this should be reported textually (currently only says "SATISFIABLE", see above example)

@AntoniusW I just noticed that this is still tagged as pending my review - have the aforementioned changes been made, i.e. can we move this forward?

AntoniusW committed:
Small test improvement.

AntoniusW committed:
Move WeakConstraintsManager into DefaultSolver.

- OptimizingSolver directly uses WeakConstraintsManager from DefaultSolver.
- Small polish on WeakConstraintRecorder.

AntoniusW committed:
Enable optimization based on command-line, also accept upper bound, and print valuation of answer sets if weak constraints are present.

- CommandLineParser and SystemConfig handle options enabling optimization and maximum upper bound.
- WeakConstraintsManager interface introduced.
- WeakConstraintsManagerForBoundedOptimality for branch-and-bound search extracted.
- WeakConstraintsManagerForUnboundedEnumeration for computing weights/levels of answer sets of programs with weak constraints.
- WeightAtLevelsManager accepts maximum upper bound from comma-seperated weight@level pairs.
- DefaultSolver adds weight/levels to answer sets if weak constraints are present.
- OptimizingSolver handles maximum upper bound if specified.
- SolverFactory returns OptimizingSolver based on flag, not presence of weak constraints in input.
- Added initialization tests to WeightAtLevelsManagerTest.
- Added to TestUtils methods for regression testing the OptimizingSolver, and using them in WeakConstraintsTests.

AntoniusW committed:
Merge branch 'master' into weak_constraints

# Conflicts:
#	alpha-api/src/main/java/at/ac/tuwien/kr/alpha/api/config/SystemConfig.java
#	alpha-cli-app/src/main/java/at/ac/tuwien/kr/alpha/Main.java
#	alpha-cli-app/src/main/java/at/ac/tuwien/kr/alpha/app/config/CommandLineParser.java
#	alpha-commons/src/main/java/at/ac/tuwien/kr/alpha/commons/BasicAnswerSet.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/atoms/WeakConstraintAtom.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/common/WeightedAnswerSet.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/grounder/Grounder.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/grounder/NaiveGrounder.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/grounder/NoGoodGenerator.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/grounder/WeakConstraintRecorder.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/parser/ParseTreeVisitor.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/programs/AnalyzedProgram.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/programs/InternalProgram.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/programs/transformation/IntervalTermToIntervalAtom.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/programs/transformation/WeakConstraintNormalization.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/rules/WeakConstraint.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/solver/AtomCallbackManager.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/solver/DefaultSolver.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/solver/OptimizingSolver.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/solver/optimization/WeakConstraintAtomCallback.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/solver/optimization/WeakConstraintsManager.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/solver/optimization/WeakConstraintsManagerForBoundedOptimality.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/solver/optimization/WeakConstraintsManagerForUnboundedEnumeration.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/solver/optimization/WeightAtLevelsManager.java
#	alpha-core/src/test/java/at/ac/tuwien/kr/alpha/core/WeightedAnswerSetTest.java
#	alpha-core/src/test/java/at/ac/tuwien/kr/alpha/core/grounder/ChoiceGrounder.java
#	alpha-core/src/test/java/at/ac/tuwien/kr/alpha/core/grounder/DummyGrounder.java
#	alpha-core/src/test/java/at/ac/tuwien/kr/alpha/core/solver/WeakConstraintsTests.java
#	alpha-core/src/test/java/at/ac/tuwien/kr/alpha/core/solver/optimization/WeightAtLevelsManagerTest.java
#	alpha-core/src/test/java/at/ac/tuwien/kr/alpha/core/test/util/TestUtils.java
#	src/main/java/at/ac/tuwien/kr/alpha/api/Alpha.java
#	src/main/java/at/ac/tuwien/kr/alpha/common/AnswerSet.java
#	src/main/java/at/ac/tuwien/kr/alpha/grounder/transformation/aggregates/AggregateOperatorNormalization.java

AntoniusW committed:
Fix merge.

AntoniusW commented:
> 
> @AntoniusW I just noticed that this is still tagged as pending my review - have the aforementioned changes been made, i.e. can we move this forward?

@madmike200590, I think it is ready now. The functionality to not run optimization but still get valuations of answer sets should work now. An (exclusive) upper bound can also be set manually. Furthermore, current master has been merged and this branch now follows the modularization introduced in master. The code is ready for another round of reviews (and subsequent merging).

AntoniusW committed:
Merge branch 'master' into weak_constraints

# Conflicts:
#	alpha-cli-app/src/main/java/at/ac/tuwien/kr/alpha/Main.java
#	alpha-cli-app/src/main/java/at/ac/tuwien/kr/alpha/app/config/CommandLineParser.java
#	alpha-commons/src/main/java/at/ac/tuwien/kr/alpha/commons/programs/ASPCore2ProgramImpl.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/grounder/NaiveGrounder.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/grounder/NoGoodGenerator.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/parser/ParseTreeVisitor.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/programs/InternalProgram.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/programs/NormalProgramImpl.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/programs/atoms/WeakConstraintAtom.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/programs/transformation/ArithmeticTermsRewriting.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/programs/transformation/IntervalTermToIntervalAtom.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/programs/transformation/VariableEqualityRemoval.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/programs/transformation/aggregates/AggregateOperatorNormalization.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/programs/transformation/aggregates/encoders/AbstractAggregateEncoder.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/programs/transformation/aggregates/encoders/StringtemplateBasedAggregateEncoder.java
#	alpha-core/src/main/java/at/ac/tuwien/kr/alpha/core/solver/DefaultSolver.java
#	alpha-core/src/test/java/at/ac/tuwien/kr/alpha/core/grounder/ChoiceGrounder.java
#	alpha-core/src/test/java/at/ac/tuwien/kr/alpha/core/solver/SolverTests.java
#	alpha-core/src/test/java/at/ac/tuwien/kr/alpha/core/test/util/TestUtils.java
#	alpha-solver/src/main/java/at/ac/tuwien/kr/alpha/api/impl/AlphaImpl.java
#	alpha-solver/src/test/java/at/ac/tuwien/kr/alpha/api/impl/AlphaImplTest.java

AntoniusW committed:
Fix merge master.

AntoniusW committed:
Correctly report if optimum answer set was found.

- Add Solver::didExhaustSearchSpace method, realize it in Solver implementations and use it in Main.

madmike200590 commented on alpha-commons/src/main/java/at/ac/tuwien/kr/alpha/commons/WeightedAnswerSet.java:
```diff
@@ -0,0 +1,154 @@
+	 * @param weightAtLevels
+	 * @return a TreeMap containing
+	 */
+	public static TreeMap<Integer, Integer> weightPerLevelFromString(String weightAtLevels) {
```
This is the wrong place to parse command-line arguments. The parsing of this string should happen in `CommandLineParser`. Internally, weights per level should only be passed around as some kind of `Map`s.

AntoniusW replied:
This is a good catch and I was considering different options for where to put this code. The main problem is, that the same functionality is needed in `WeakConstraintsTests` in the tests of the core module. Relying on code from the cli-app module there felt like a bad idea and the other option would be having the code duplicated. What would you suggest for this kind of dependency?

madmike200590 reviewed (changes requested):
Looks good from my initial tests!
One small thing to fix - commandline argument parsing for weight/level pairs is a bit "over the place"
